[Filters] Make Filter and FilterEffects create FilterStyles

shallawa opened the pull request:
#### bd6357f66d63331d03f57403dfa38a194a552ea4
<pre>
[Filters] Make Filter and FilterEffects create FilterStyles
<a href="https://bugs.webkit.org/show_bug.cgi?id=248318">https://bugs.webkit.org/show_bug.cgi?id=248318</a>
rdar://102651202

Reviewed by Simon Fraser.

A FilterStyle is a structure composed of a GraphicsStyle and the geometry of the
FilterEffect. A vector of FilterStyles will be used to apply a chain of FilterEffects
to the GraphicsContext. The FilterStyle can only be created by effects which takes
one and only one input.

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterRenderingModes const):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGStyle):
(WebCore::GraphicsContextCG::didUpdateState):
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterRenderingModes const):
(WebCore::FEColorMatrix::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FEColorMatrix.h:
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.cpp:
(WebCore::FEComponentTransfer::supportedFilterRenderingModes const):
* Source/WebCore/platform/graphics/filters/FEDropShadow.cpp:
(WebCore::FEDropShadow::supportedFilterRenderingModes const):
(WebCore::FEDropShadow::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FEDropShadow.h:
* Source/WebCore/platform/graphics/filters/FEGaussianBlur.cpp:
(WebCore::FEGaussianBlur::supportedFilterRenderingModes const):
(WebCore::FEGaussianBlur::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FEGaussianBlur.h:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::createFilterStyles const):
* Source/WebCore/platform/graphics/filters/Filter.h:
* Source/WebCore/platform/graphics/filters/FilterEffect.cpp:
(WebCore::FilterEffect::createFilterStyles const):
(WebCore::FilterEffect::createFilterStyle const):
* Source/WebCore/platform/graphics/filters/FilterEffect.h:
(WebCore::FilterEffect::numberOfImageInputs const):
(WebCore::FilterEffect::numberOfEffectInputs const):
(WebCore::FilterEffect::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FilterFunction.h:
(WebCore::FilterFunction::createFilterStyles const):
* Source/WebCore/platform/graphics/filters/FilterStyle.h: Added.
* Source/WebCore/platform/graphics/filters/SourceGraphic.cpp:
(WebCore::SourceGraphic::supportedFilterRenderingModes const):
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::createFilterStyles const):
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::apply):
(WebCore::SVGFilter::createFilterStyles const):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
</pre>

smfr commented on Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
```diff
@@ -116,6 +116,10 @@ OptionSet<FilterRenderingMode> FEColorMatrix::supportedFilterRenderingModes() co
+#endif
+#if USE(GRAPHICS_CONTEXT_FILTERS)
+    if (m_type == FECOLORMATRIX_TYPE_MATRIX)
+        modes = modes | FilterRenderingMode::GraphicsContext;
```
This needs the OS-version based `#ifdef` so that we don't use it on downlevels.

shallawa replied:
`#if HAVE(CGSTYLE_COLORMATRIX_BLUR)` is used instead.

smfr commented on Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
```diff
@@ -134,6 +138,13 @@ std::unique_ptr<FilterEffectApplier> FEColorMatrix::createSoftwareApplier() cons
     return FilterEffectApplier::create<FEColorMatrixSoftwareApplier>(*this);
 }
 
+std::optional<GraphicsStyle> FEColorMatrix::createGraphicsStyle(const Filter&) const
```
I don't think we'll ever get here for invert, opacity, brightness or contrast unless you also take bits of https://bugs.webkit.org/show_bug.cgi?id=245111

Also, I'd prefer that we write this in way where we can call functions on `ColorMatrix<>`. We should take parts of https://bugs.webkit.org/show_bug.cgi?id=245111 first.

shallawa replied:
I agree but can we do this is a separate patch after landing https://bugs.webkit.org/show_bug.cgi?id=245111?

smfr replied:
OK

smfr commented on Source/WebCore/platform/graphics/filters/FEGaussianBlur.cpp:
```diff
@@ -140,11 +140,27 @@ bool FEGaussianBlur::resultIsAlphaImage(const FilterImageVector& inputs) const
+{
+    OptionSet<FilterRenderingMode> modes = FilterRenderingMode::Software;
+#if USE(GRAPHICS_CONTEXT_FILTERS)
+    if (m_stdX == m_stdY)
```
This requires the new platform version-based #ifdef.

shallawa replied:
`#if HAVE(CGSTYLE_COLORMATRIX_BLUR)` is used instead.

smfr commented on Source/WebCore/platform/graphics/filters/FEGaussianBlur.cpp:
```diff
@@ -140,11 +140,27 @@ bool FEGaussianBlur::resultIsAlphaImage(const FilterImageVector& inputs) const
 
+std::optional<GraphicsStyle> FEGaussianBlur::createGraphicsStyle(const Filter& filter) const
+{
+    auto radius = 2 * filter.resolvedSize({ m_stdX, m_stdY });
```
Are you sure this relationship between the stdDeviation and the radius is correct?

shallawa replied:
This is almost correct. But we can use `calculateUnscaledKernelSize()` which will multiply the resolved size by `3 / 4.f * sqrtf(2 * piFloat)` which is equal to 1.8799. We call this function when calculating the imageRect of the `FEGaussianBlur` and `FEDropShadow`.

I borrowed the above calculation from `FEDropShadowSoftwareApplier::apply()`.

smfr commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -1087,25 +1086,39 @@ static void setCGStyle(CGContextRef context, const std::optional<GraphicsStyle>&
+#if HAVE(CGSTYLE_CREATE_SHADOW2)
             return adoptCF(CGStyleCreateShadow2(dropShadow.offset, dropShadow.radius.width(), cachedCGColor(dropShadow.color).get()));
+#else
+            UNUSED_PARAM(dropShadow);
```
Should we assert that we never get here?

shallawa replied:
Done.

smfr commented on Source/WebCore/platform/graphics/filters/FilterStyle.h:
```diff
@@ -0,0 +1,41 @@
+namespace WebCore {
+
+struct FilterStyle {
+    std::optional<GraphicsStyle> style;
```
What does having `style` be unset represent?

shallawa replied:
It is used as an input for the SourceGraphics effect. See `Filter::createFilterStyles()`.

smfr commented on Source/WebCore/platform/graphics/filters/SourceGraphic.cpp:
```diff
@@ -46,6 +46,9 @@ OptionSet<FilterRenderingMode> SourceGraphic::supportedFilterRenderingModes() co
     modes = modes | FilterRenderingMode::Accelerated;
+#endif
+#if USE(GRAPHICS_CONTEXT_FILTERS)
+    modes = modes | FilterRenderingMode::GraphicsContext;
```
I find using `OptionSet<>::add()` clearer.

shallawa replied:
Done.

shallawa committed:
[Filters] Make Filter and FilterEffects create FilterStyles
https://bugs.webkit.org/show_bug.cgi?id=248318
rdar://102651202

Reviewed by Simon Fraser.

A FilterStyle is a structure composed of a GraphicsStyle and the geometry of the
FilterEffect. A vector of FilterStyles will be used to apply a chain of FilterEffects
to the GraphicsContext. The FilterStyle can only be created by effects which takes
one and only one input.

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterRenderingModes const):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGStyle):
(WebCore::GraphicsContextCG::didUpdateState):
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterRenderingModes const):
(WebCore::FEColorMatrix::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FEColorMatrix.h:
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.cpp:
(WebCore::FEComponentTransfer::supportedFilterRenderingModes const):
* Source/WebCore/platform/graphics/filters/FEDropShadow.cpp:
(WebCore::FEDropShadow::supportedFilterRenderingModes const):
(WebCore::FEDropShadow::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FEDropShadow.h:
* Source/WebCore/platform/graphics/filters/FEGaussianBlur.cpp:
(WebCore::FEGaussianBlur::supportedFilterRenderingModes const):
(WebCore::FEGaussianBlur::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FEGaussianBlur.h:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::createFilterStyles const):
* Source/WebCore/platform/graphics/filters/Filter.h:
* Source/WebCore/platform/graphics/filters/FilterEffect.cpp:
(WebCore::FilterEffect::createFilterStyles const):
(WebCore::FilterEffect::createFilterStyle const):
* Source/WebCore/platform/graphics/filters/FilterEffect.h:
(WebCore::FilterEffect::numberOfImageInputs const):
(WebCore::FilterEffect::numberOfEffectInputs const):
(WebCore::FilterEffect::createGraphicsStyle const):
* Source/WebCore/platform/graphics/filters/FilterFunction.h:
(WebCore::FilterFunction::createFilterStyles const):
* Source/WebCore/platform/graphics/filters/FilterStyle.h: Added.
* Source/WebCore/platform/graphics/filters/SourceGraphic.cpp:
(WebCore::SourceGraphic::supportedFilterRenderingModes const):
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::createFilterStyles const):
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::apply):
(WebCore::SVGFilter::createFilterStyles const):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:

Canonical link: https://commits.webkit.org/257045@main
